File: C/linear-search.c
#define CRT_SECURE_NO_WARNINGS

#include <stdio.h>

int linearSearch(int arr[], int arrLen, int searchFor) {
    //  Linear search: Iterates through an array until the value is found or the array is fully traversed.
    //  Time Complexity: Best - O(1) Average - O(n) Worst - O(n)
    //  NOTE: Best case is when the value is in the first index.

    //  Declare i as type int
    int i;

    //  Iterate over array, comparing value at current index to value searched for
    for (i = 0; i < arrLen; i++) {

        if (arr[i] == searchFor) {
            //  Return the index if the value is found
            return i;
        }
    }

    //  Return -1 if the value is not found and the array is fully traversed
    return -1;
}

//  Driver program for testing the linear search
int main() {
    int arr[100], arrLen, i, searchFor, result;

    printf("Enter number of elements in the array: ");
    scanf("%d", &arrLen);

    printf("Enter %d integers\n\r", arrLen);

    for (i = 0; i < arrLen; i++) {
        scanf("%d", &arr[i]);
    }

    printf("Enter the value you'd like to search for: ");
    scanf("%d", &searchFor);

    result = linearSearch(arr, arrLen, searchFor);

    if (result != -1) {
        printf("%d was found in position %d (index %d).\n\r", searchFor, result + 1, result);
    } else {
        printf("%d was not found in the array.\n\r", searchFor);
    }

    return 0;
}

File: Python/linear-search.py
def linear_search(input_arr, search_for):
    # Linear search: Iterates through an array until the value is found or the array is fully traversed.
    # Time Complexity: Best - O(1) Average - O(n) Worst - O(n)
    # NOTE: Best case is when the value is in the first index.

    # For loop through each index of the length of the array
    for i in range(0, len(input_arr), 1):

        # If the value at the current index matches the search for value, return the index
        if input_arr[i] == search_for:
            return i

    # If the entire array is traversed and the value has not been found, return -1
    return -1


#   Driver code
if __name__ == "__main__":
    my_array = [1, 564, 1123, 9999, 5641, 1234, 5, 14, 89, 445]

    print(my_array)

    index = linear_search(my_array, 9999)

    if index != -1:
        print("Value found at index " + str(index))
    else:
        print("Value not found.")
